File: server/internal/ddb/sharedtxns.go
package ddb

import (
	"encoding/json"
	"fmt"

	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/nabeken/aws-go-dynamodb/attributes"
	"github.com/nabeken/aws-go-dynamodb/table"
	"github.com/nabeken/aws-go-dynamodb/table/option"
	"github.com/saifahn/expenseus/internal/app"
)

const (
	sharedTxnKeyPrefix  = "txn.shared"
	sharedTxnEntityType = "sharedTransaction"
	unsettledFlagTrue   = "X"
)

type SharedTxnItem struct {
	PK           string   `json:"PK"`
	SK           string   `json:"SK"`
	GSI1PK       string   `json:"GSI1PK"`
	GSI1SK       string   `json:"GSI1SK"`
	EntityType   string   `json:"EntityType"`
	ID           string   `json:"ID"`
	Date         int64    `json:"Date"`
	Amount       int64    `json:"Amount"`
	Location     string   `json:"Location"`
	Tracker      string   `json:"Tracker"`
	Category     string   `json:"Category"`
	Participants []string `json:"Participants"`
	Payer        string   `json:"Payer"`
	Unsettled    string   `json:"Unsettled,omitempty"`
	Details      string   `json:"Details"`
	SplitJSON    string   `json:"Split"`
}

type SettleTxnInput struct {
	ID           string
	TrackerID    string
	Participants []string
}

type SharedTxnsRepository interface {
	Create(txn app.SharedTransaction) error
	GetFromTracker(trackerID string) ([]SharedTxnItem, error)
	GetFromTrackerBetweenDates(trackerID string, from, to int64) ([]SharedTxnItem, error)
	GetByUserBetweenDates(userID string, from, to int64) ([]SharedTxnItem, error)
	GetUnsettledFromTracker(trackerID string) ([]SharedTxnItem, error)
	Update(txn app.SharedTransaction) error
	Delete(input app.DelSharedTxnInput) error
	Settle(input []SettleTxnInput) error
}

type sharedTxnsRepo struct {
	table *table.Table
}

func NewSharedTxnsRepository(t *table.Table) SharedTxnsRepository {
	t.WithHashKey(tablePrimaryKey, dynamodb.ScalarAttributeTypeS)
	t.WithRangeKey(tableSortKey, dynamodb.ScalarAttributeTypeS)
	return &sharedTxnsRepo{t}
}

func (r *sharedTxnsRepo) Create(txn app.SharedTransaction) error {
	trackerIDKey := makeTrackerIDKey(txn.Tracker)
	txnIDKey := makeSharedTxnIDKey(txn.ID)
	txnDateKey := makeSharedTxnDateIDKey(txn)

	txnItem := SharedTxnItem{
		SK:           txnIDKey,
		GSI1SK:       txnDateKey,
		EntityType:   sharedTxnEntityType,
		ID:           txn.ID,
		Category:     txn.Category,
		Tracker:      txn.Tracker,
		Participants: txn.Participants,
		Date:         txn.Date,
		Amount:       txn.Amount,
		Location:     txn.Location,
		Payer:        txn.Payer,
		Details:      txn.Details,
	}

	if txn.Unsettled {
		txnItem.Unsettled = unsettledFlagTrue
	}
	// store the split map as a JSON string
	if txn.Split != nil {
		splitJSON, err := json.Marshal(txn.Split)
		if err != nil {
			return err
		}
		txnItem.SplitJSON = string(splitJSON)
	}

	for _, p := range txn.Participants {
		userIDKey := makeUserIDKey(p)
		txnItem.PK = userIDKey
		txnItem.GSI1PK = userIDKey
		err := r.table.PutItem(txnItem)
		if err != nil {
			return err
		}
	}

	txnItem.GSI1PK = trackerIDKey
	txnItem.PK = trackerIDKey
	err := r.table.PutItem(txnItem)
	return err
}

func (r *sharedTxnsRepo) GetFromTracker(trackerID string) ([]SharedTxnItem, error) {
	trackerIDKey := makeTrackerIDKey(trackerID)

	options := []option.QueryInput{
		option.Index(gsi1Name),
		option.QueryExpressionAttributeName(gsi1PrimaryKey, "#GSI1PK"),
		option.QueryExpressionAttributeName(gsi1SortKey, "#GSI1SK"),
		option.QueryExpressionAttributeValue(":trackerIDKey", attributes.String(trackerIDKey)),
		option.QueryExpressionAttributeValue(":sharedTxnKeyPrefix", attributes.String(sharedTxnKeyPrefix)),
		// sort descending
		option.Reverse(),
		option.QueryKeyConditionExpression("#GSI1PK = :trackerIDKey and begins_with(#GSI1SK, :sharedTxnKeyPrefix)"),
	}

	var items []SharedTxnItem

	_, err := r.table.Query(&items, options...)
	if err != nil {
		return nil, err
	}

	return items, nil
}

func (r *sharedTxnsRepo) GetFromTrackerBetweenDates(trackerID string, from, to int64) ([]SharedTxnItem, error) {
	trackerIDKey := makeTrackerIDKey(trackerID)
	txnDateFromKey := makeSharedTxnDateKey(from)
	txnDateToKey := makeSharedTxnDateKey(to)

	options := []option.QueryInput{
		option.Index(gsi1Name),
		option.QueryExpressionAttributeName(gsi1PrimaryKey, "#GSI1PK"),
		option.QueryExpressionAttributeName(gsi1SortKey, "#GSI1SK"),
		option.QueryExpressionAttributeValue(":trackerKey", attributes.String(trackerIDKey)),
		option.QueryExpressionAttributeValue(":txnDateFromKey", attributes.String(txnDateFromKey)),
		option.QueryExpressionAttributeValue(":txnDateToKey", attributes.String(txnDateToKey)),
		option.QueryKeyConditionExpression("#GSI1PK = :trackerKey and #GSI1SK BETWEEN :txnDateFromKey AND :txnDateToKey"),
	}

	var items []SharedTxnItem
	_, err := r.table.Query(&items, options...)
	if err != nil {
		return nil, err
	}

	return items, nil
}

func (r *sharedTxnsRepo) GetByUserBetweenDates(userID string, from, to int64) ([]SharedTxnItem, error) {
	userIDKey := makeUserIDKey(userID)
	txnDateFromKey := makeSharedTxnDateKey(from)
	txnDateToKey := makeSharedTxnDateKey(to)

	options := []option.QueryInput{
		option.Index(gsi1Name),
		option.QueryExpressionAttributeName(gsi1PrimaryKey, "#GSI1PK"),
		option.QueryExpressionAttributeName(gsi1SortKey, "#GSI1SK"),
		option.QueryExpressionAttributeValue(":userKey", attributes.String(userIDKey)),
		option.QueryExpressionAttributeValue(":txnDateFromKey", attributes.String(txnDateFromKey)),
		option.QueryExpressionAttributeValue(":txnDateToKey", attributes.String(txnDateToKey)),
		option.QueryKeyConditionExpression("#GSI1PK = :userKey and #GSI1SK BETWEEN :txnDateFromKey AND :txnDateToKey"),
	}

	var items []SharedTxnItem
	_, err := r.table.Query(&items, options...)
	if err != nil {
		return nil, err
	}

	return items, nil
}

func (r *sharedTxnsRepo) GetUnsettledFromTracker(trackerID string) ([]SharedTxnItem, error) {
	trackerIDKey := makeTrackerIDKey(trackerID)

	options := []option.QueryInput{
		option.Index(unsettledTxnsIndexName),
		option.QueryExpressionAttributeName(unsettledTxnsIndexPK, "#unsettledPK"),
		option.QueryExpressionAttributeName(unsettledTxnsIndexSK, "#unsettledSK"),
		option.QueryExpressionAttributeValue(":trackerID", attributes.String(trackerIDKey)),
		option.QueryExpressionAttributeValue(":true", attributes.String(unsettledFlagTrue)),
		option.QueryKeyConditionExpression("#unsettledPK = :trackerID and #unsettledSK = :true"),
	}

	var items []SharedTxnItem

	_, err := r.table.Query(&items, options...)
	if err != nil {
		return nil, err
	}

	return items, nil
}

func (r *sharedTxnsRepo) Update(txn app.SharedTransaction) error {
	trackerIDKey := makeTrackerIDKey(txn.Tracker)
	txnIDKey := makeSharedTxnIDKey(txn.ID)
	txnDateIDKey := makeSharedTxnDateIDKey(txn)

	txnItem := SharedTxnItem{
		SK:           txnIDKey,
		GSI1SK:       txnDateIDKey,
		EntityType:   sharedTxnEntityType,
		ID:           txn.ID,
		Category:     txn.Category,
		Tracker:      txn.Tracker,
		Participants: txn.Participants,
		Date:         txn.Date,
		Amount:       txn.Amount,
		Location:     txn.Location,
		Payer:        txn.Payer,
		Details:      txn.Details,
	}

	if txn.Unsettled {
		txnItem.Unsettled = unsettledFlagTrue
	}
	// store the split map as a JSON string
	if txn.Split != nil {
		splitJSON, err := json.Marshal(txn.Split)
		if err != nil {
			return err
		}
		txnItem.SplitJSON = string(splitJSON)
	}

	for _, p := range txn.Participants {
		userIDKey := makeUserIDKey(p)
		txnItem.PK = userIDKey
		txnItem.GSI1PK = userIDKey
		err := r.table.PutItem(txnItem, option.PutCondition("attribute_exists(SK)"))
		if err != nil {
			return attrNotExistsOrErr(err)
		}
	}

	txnItem.PK = trackerIDKey
	txnItem.GSI1PK = trackerIDKey
	err := r.table.PutItem(txnItem, option.PutCondition("attribute_exists(SK)"))
	return attrNotExistsOrErr(err)
}

func (r *sharedTxnsRepo) Delete(input app.DelSharedTxnInput) error {
	trackerIDKey := makeTrackerIDKey(input.Tracker)
	txnIDKey := makeSharedTxnIDKey(input.TxnID)

	for _, p := range input.Participants {
		userIDKey := makeUserIDKey(p)
		err := r.table.DeleteItem(attributes.String(userIDKey), attributes.String(txnIDKey))
		if err != nil {
			return err
		}
	}

	return r.table.DeleteItem(attributes.String(trackerIDKey), attributes.String(txnIDKey))
}

// Settle takes a slice of SettleTxnInputs and removes the "Unsettled"
// attribute from the database items that correspond to the transactions
func (r *sharedTxnsRepo) Settle(txns []SettleTxnInput) error {
	var updateOpts []option.UpdateItemInput
	updateOpts = append(updateOpts, option.UpdateExpressionAttributeName("Unsettled", "#unsettled"))
	updateOpts = append(updateOpts, option.UpdateExpression("REMOVE #unsettled"))

	for _, t := range txns {
		trackerIDKey := makeTrackerIDKey(t.TrackerID)
		txnIDKey := makeSharedTxnIDKey(t.ID)
		_, err := r.table.UpdateItem(
			attributes.String(trackerIDKey),
			attributes.String(txnIDKey),
			updateOpts...,
		)
		if err != nil {
			return err
		}

		for _, u := range t.Participants {
			userKey := makeUserIDKey(u)
			_, err := r.table.UpdateItem(
				attributes.String(userKey),
				attributes.String(txnIDKey),
				updateOpts...,
			)
			if err != nil {
				return err
			}
		}
	}
	return nil
}

func makeSharedTxnIDKey(id string) string {
	return fmt.Sprintf("%s#%s", sharedTxnKeyPrefix, id)
}

func makeSharedTxnDateIDKey(txn app.SharedTransaction) string {
	return fmt.Sprintf("%s#%d#%s", sharedTxnKeyPrefix, txn.Date, txn.ID)
}

func makeSharedTxnDateKey(date int64) string {
	return fmt.Sprintf("%s#%d", sharedTxnKeyPrefix, date)
}

File: client/ddb/sharedTxns.ts
import { ConditionalCheckFailedException } from '@aws-sdk/client-dynamodb';
import {
  DeleteCommand,
  PutCommand,
  QueryCommand,
  UpdateCommand,
} from '@aws-sdk/lib-dynamodb';
import { SubcategoryKey } from 'data/categories';
import { CreateSharedTxnPayload } from 'pages/api/v1/trackers/[trackerId]/transactions';
import { UpdateSharedTxnPayload } from 'pages/api/v1/trackers/[trackerId]/transactions/[txnId]';
import { monotonicFactory } from 'ulid';
import { ItemDoesNotExistError } from './errors';
import {
  DDBWithConfig,
  gsi1PartitionKey,
  gsi1SortKey,
  gsi1Name,
  tablePartitionKey,
  tableSortKey,
  unsettledTxnsIndexName,
  unsettledTxnsIndexPK,
  unsettledTxnsIndexSK,
} from './schema';
import { makeTrackerIdKey } from './trackers';
import { makeUserIdKey } from './users';

const sharedTxnKeyPrefix = 'txn.shared',
  sharedTxnEntityType = 'sharedTransaction',
  unsettledFlagTrue = 'X';

const makeSharedTxnIdKey = (id: string) => `${sharedTxnKeyPrefix}#${id}`;
const makeSharedTxnDateKey = (date: number) =>
  `${sharedTxnKeyPrefix}#${date.toString()}`;
const makeSharedTxnDateIdKey = ({ date, id }: { date: number; id: string }) =>
  `${sharedTxnKeyPrefix}#${date}#${id}`;

export type SharedTxnItem = {
  [tablePartitionKey]: string;
  [tableSortKey]: string;
  [gsi1PartitionKey]: string;
  [gsi1SortKey]: string;
  EntityType: typeof sharedTxnEntityType;
  ID: string;
  Date: number;
  Amount: number;
  Location: string;
  Tracker: string;
  Category: SubcategoryKey;
  Participants: string[];
  Payer: string;
  Details: string;
  Unsettled?: typeof unsettledFlagTrue;
  SplitJSON?: string;
};

export type SharedTxn = {
  id: string;
  date: number;
  amount: number;
  location: string;
  tracker: string;
  category: SubcategoryKey;
  participants: string[];
  payer: string;
  details: string;
  unsettled?: boolean;
  split?: {
    [k: string]: number;
  };
};

const ulid = monotonicFactory();

export function makeSharedTxnRepository({ ddb, tableName }: DDBWithConfig) {
  /**
   * Creates a shared transaction based on the given input.
   */
  async function createSharedTxn(txn: CreateSharedTxnPayload) {
    const txnId = ulid(txn.date);
    const txnIdKey = makeSharedTxnIdKey(txnId);
    const txnDateIdKey = makeSharedTxnDateIdKey({ id: txnId, date: txn.date });

    // TODO: add type for this as well?
    const item = {
      [tableSortKey]: txnIdKey,
      [gsi1SortKey]: txnDateIdKey,
      EntityType: sharedTxnEntityType,
      ID: txnId,
      Category: txn.category,
      Tracker: txn.tracker,
      Participants: txn.participants,
      Date: txn.date,
      Amount: txn.amount,
      Location: txn.location,
      Payer: txn.payer,
      Details: txn.details,
      ...(txn.unsettled && { Unsettled: unsettledFlagTrue }),
      // storing a JS object using the ddb Map type would be better, but the original
      // implementation used a string so this is kept for compatibility
      ...(txn.split && { SplitJSON: JSON.stringify(txn.split) }),
    };

    // store the representation of the tracker under each user so trackers can
    // be retrieved for them
    for (const user of txn.participants) {
      const userIdKey = makeUserIdKey(user);
      const uItem = {
        ...item,
        [tablePartitionKey]: userIdKey,
        [gsi1PartitionKey]: userIdKey,
      };
      await ddb.send(
        new PutCommand({
          TableName: tableName,
          Item: uItem,
        }),
      );
    }

    const trackerIdKey = makeTrackerIdKey(txn.tracker);
    const trackerItem = {
      ...item,
      [tablePartitionKey]: trackerIdKey,
      [gsi1PartitionKey]: trackerIdKey,
    };
    await ddb.send(
      new PutCommand({
        TableName: tableName,
        Item: trackerItem,
      }),
    );
  }

  /**
   * Updates a shared transaction based on the given input.
   */
  async function updateSharedTxn(txn: UpdateSharedTxnPayload) {
    // TODO: update the 2nd arg type so it has to have an ID
    const txnIdKey = makeSharedTxnIdKey(txn.id);
    const txnDateIdKey = makeSharedTxnDateIdKey(txn);

    const item = {
      [tableSortKey]: txnIdKey,
      [gsi1SortKey]: txnDateIdKey,
      EntityType: sharedTxnEntityType,
      ID: txn.id,
      Category: txn.category,
      Tracker: txn.tracker,
      Participants: txn.participants,
      Date: txn.date,
      Amount: txn.amount,
      Location: txn.location,
      Payer: txn.payer,
      Details: txn.details,
      ...(txn.unsettled && { Unsettled: unsettledFlagTrue }),
      ...(txn.split && { SplitJSON: JSON.stringify(txn.split) }),
    };

    for (const user of txn.participants) {
      const userIdKey = makeUserIdKey(user);
      const uItem = {
        ...item,
        [tablePartitionKey]: userIdKey,
        [gsi1PartitionKey]: userIdKey,
      };
      try {
        await ddb.send(
          new PutCommand({
            TableName: tableName,
            Item: uItem,
            ExpressionAttributeNames: {
              '#SK': tableSortKey,
            },
            ConditionExpression: 'attribute_exists(#SK)',
          }),
        );
      } catch (err) {
        if (err instanceof ConditionalCheckFailedException) {
          throw new ItemDoesNotExistError('shared txn does not exist');
        }
        throw err;
      }
    }

    const trackerIdKey = makeTrackerIdKey(txn.tracker);
    const trackerItem = {
      ...item,
      [tablePartitionKey]: trackerIdKey,
      [gsi1PartitionKey]: trackerIdKey,
    };
    try {
      await ddb.send(
        new PutCommand({
          TableName: tableName,
          Item: trackerItem,
          ExpressionAttributeNames: {
            '#SK': tableSortKey,
          },
          ConditionExpression: 'attribute_exists(#SK)',
        }),
      );
    } catch (err) {
      if (err instanceof ConditionalCheckFailedException) {
        throw new ItemDoesNotExistError('shared txn does not exist');
      }
      throw err;
    }
  }

  type DeleteSharedTxnInput = {
    tracker: string;
    txnId: string;
    participants: string[];
  };

  /**
   * Deletes a shared transaction based on the given input
   */
  async function deleteSharedTxn(input: DeleteSharedTxnInput) {
    const txnIdKey = makeSharedTxnIdKey(input.txnId);

    for (const user of input.participants) {
      const userIdKey = makeUserIdKey(user);
      await ddb.send(
        new DeleteCommand({
          TableName: tableName,
          Key: {
            [tablePartitionKey]: userIdKey,
            [tableSortKey]: txnIdKey,
          },
        }),
      );
    }

    const trackerIdKey = makeTrackerIdKey(input.tracker);
    await ddb.send(
      new DeleteCommand({
        TableName: tableName,
        Key: {
          [tablePartitionKey]: trackerIdKey,
          [tableSortKey]: txnIdKey,
        },
      }),
    );
  }

  /**
   * Retrieves all shared transactions from the tracker with the given tracker ID.
   */
  async function getTxnsByTracker(trackerId: string) {
    const trackerIdKey = makeTrackerIdKey(trackerId);

    const result = await ddb.send(
      new QueryCommand({
        TableName: tableName,
        IndexName: gsi1Name,
        ExpressionAttributeNames: {
          '#GSI1PK': gsi1PartitionKey,
          '#GSI1SK': gsi1SortKey,
        },
        ExpressionAttributeValues: {
          ':trackerIdKey': trackerIdKey,
          ':sharedTxnKeyPrefix': sharedTxnKeyPrefix,
        },
        // return in descending order
        ScanIndexForward: false,
        KeyConditionExpression:
          '#GSI1PK = :trackerIdKey and begins_with(#GSI1SK, :sharedTxnKeyPrefix)',
      }),
    );

    return (result.Items as SharedTxnItem[]) ?? [];
  }

  type ByTrackerBetweenDatesInput = {
    tracker: string;
    from: number;
    to: number;
  };
  /**
   * Retrieves transactions between the given dates for the given tracker ID.
   */
  async function getTxnsByTrackerBetweenDates({
    tracker,
    from,
    to,
  }: ByTrackerBetweenDatesInput) {
    const trackerIdKey = makeTrackerIdKey(tracker);
    const txnDateFromKey = makeSharedTxnDateKey(from);
    const txnDateToKey = makeSharedTxnDateKey(to + 1); // to be inclusive for the date

    const results = await ddb.send(
      new QueryCommand({
        TableName: tableName,
        IndexName: gsi1Name,
        ExpressionAttributeNames: {
          '#GSI1PK': gsi1PartitionKey,
          '#GSI1SK': gsi1SortKey,
        },
        ExpressionAttributeValues: {
          ':trackerKey': trackerIdKey,
          ':txnDateFromKey': txnDateFromKey,
          ':txnDateToKey': txnDateToKey,
        },
        KeyConditionExpression:
          '#GSI1PK = :trackerKey and #GSI1SK BETWEEN :txnDateFromKey AND :txnDateToKey',
      }),
    );

    return (results.Items as SharedTxnItem[]) ?? [];
  }

  type ByUserBetweenDatesInput = {
    user: string;
    from: number;
    to: number;
  };
  /**
   * Retrieves shared transactions between the given dates for the given user ID.
   */
  async function getSharedTxnsByUserBetweenDates({
    user,
    from,
    to,
  }: ByUserBetweenDatesInput) {
    const userIdKey = makeUserIdKey(user);
    const txnDateFromKey = makeSharedTxnDateKey(from);
    const txnDateToKey = makeSharedTxnDateKey(to);

    const results = await ddb.send(
      new QueryCommand({
        TableName: tableName,
        IndexName: gsi1Name,
        ExpressionAttributeNames: {
          '#GSI1PK': gsi1PartitionKey,
          '#GSI1SK': gsi1SortKey,
        },
        ExpressionAttributeValues: {
          ':trackerKey': userIdKey,
          ':txnDateFromKey': txnDateFromKey,
          ':txnDateToKey': txnDateToKey,
        },
        KeyConditionExpression:
          '#GSI1PK = :trackerKey and #GSI1SK BETWEEN :txnDateFromKey AND :txnDateToKey',
      }),
    );

    return (results.Items as SharedTxnItem[]) ?? [];
  }

  /**
   * Retrieves unsettled transactions from the tracker with the given tracker ID.
   */
  async function getUnsettledTxnsByTracker(trackerId: string) {
    const trackerIdKey = makeTrackerIdKey(trackerId);

    const result = await ddb.send(
      new QueryCommand({
        TableName: tableName,
        IndexName: unsettledTxnsIndexName,
        ExpressionAttributeNames: {
          '#unsettledPK': unsettledTxnsIndexPK,
          '#unsettledSK': unsettledTxnsIndexSK,
        },
        ExpressionAttributeValues: {
          ':trackerIdKey': trackerIdKey,
          ':true': unsettledFlagTrue,
        },
        KeyConditionExpression:
          '#unsettledPK = :trackerIdKey and #unsettledSK = :true',
      }),
    );
    return (result.Items as SharedTxnItem[]) ?? [];
  }

  type SettleTxnInput = {
    id: string;
    trackerId: string;
    participants: string[];
  };
  /**
   * Settles transactions for the given input.
   */
  async function settleTxns(txns: SettleTxnInput[]) {
    for (const txn of txns) {
      const trackerIdKey = makeTrackerIdKey(txn.trackerId);
      const txnIdKey = makeSharedTxnIdKey(txn.id);
      await ddb.send(
        new UpdateCommand({
          TableName: tableName,
          Key: {
            [tablePartitionKey]: trackerIdKey,
            [tableSortKey]: txnIdKey,
          },
          ExpressionAttributeNames: {
            '#unsettled': 'Unsettled',
          },
          UpdateExpression: 'REMOVE #unsettled',
        }),
      );

      for (const user of txn.participants) {
        const userIdKey = makeUserIdKey(user);
        await ddb.send(
          new UpdateCommand({
            TableName: tableName,
            Key: {
              [tablePartitionKey]: userIdKey,
              [tableSortKey]: txnIdKey,
            },
            ExpressionAttributeNames: {
              '#unsettled': 'Unsettled',
            },
            UpdateExpression: 'REMOVE #unsettled',
          }),
        );
      }
    }
  }

  return {
    createSharedTxn,
    updateSharedTxn,
    deleteSharedTxn,
    getTxnsByTracker,
    getTxnsByTrackerBetweenDates,
    getSharedTxnsByUserBetweenDates,
    getUnsettledTxnsByTracker,
    settleTxns,
  };
}
